ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_matmul/device/multi_core/all_gather_matmul_op_multi_core.cpp

```diff
@@ -180,34 +195,191 @@ DatacopyParams setup_datacopy(
+    );
 
-    // Pass in the datacopy cores and sempahore address (Using optional arguments)
-    operation::ProgramWithCallbacks program_with_callbacks = ttnn::all_gather_multi_core_with_workers_helper(program, input_tensor, all_gather_output_tensor, dim, num_links, ring_size, ring_index, receiver_device_id, sender_device_id, topology, fused_op_signaler, core_grid_offset);
-    auto all_gather_override_runtime_arguments_callback = program_with_callbacks.override_runtime_arguments_callback;
+    // Matmul
+    std::optional<operation::ProgramWithCallbacks> matmul_program_with_callbacks;
+    std::optional<operation::OverrideRuntimeArgumentsCallback<Tensors>> matmul_override_runtime_arguments_callback;
+
+    std::visit([&] (const auto& config) {
+        using ProgramConfigType = std::decay_t<decltype(config)>;
+        if (std::is_same_v<ProgramConfigType, operations::matmul::MatmulMultiCoreReuseMultiCastProgramConfig>) {
+            matmul_program_with_callbacks = operations::matmul::matmul_multi_core_reuse_mcast_2d_optimized_helper(
+                program,
+                all_gather_output_tensor,
+                weight_tensor,
+                bias,
+                matmul_output_tensor,
+                bcast_batch,
+                compute_kernel_config,
+                config,
+                untilize_out,
+                matmul_fused_op_signaler
+            );
+            matmul_override_runtime_arguments_callback = matmul_program_with_callbacks->override_runtime_arguments_callback;
+        } else if (std::is_same_v<ProgramConfigType, operations::matmul::MatmulMultiCoreReuseMultiCast1DProgramConfig>) {
+            matmul_program_with_callbacks = operations::matmul::matmul_multi_core_reuse_mcast_1d_optimized_helper(
+                program,
+                all_gather_output_tensor,
+                weight_tensor,
+                bias,
+                matmul_output_tensor,
+                bcast_batch,
+                compute_kernel_config,
+                config,
+                untilize_out,
+                matmul_fused_op_signaler
+            );
+            matmul_override_runtime_arguments_callback = matmul_program_with_callbacks->override_runtime_arguments_callback;
+        } else {
+            TT_FATAL("Unsupported MatmulProgramConfig type");
+        }
+    }, program_config);
+
+    if (!matmul_program_with_callbacks.has_value()) {
+        TT_FATAL("Matmul program with callbacks not created");
+    }
+
+
+    // Datacopy
+    DatacopyParams datacopy_params;
+    if (use_datacopy) {
+        datacopy_params = setup_datacopy(
+            matmul_program_with_callbacks->program,
+            input_tensor,
+            all_gather_output_tensor,
+            datacopy_output_tensor,
+            dim,
+            num_links,
+            ring_size,
+            ring_index,
+            topology,
+            {0, 7},
```

Comment:
Would it be possible to define a constant or maybe variable instead of having a magic number of 7?